scripts/analysis/2_1_negative_net_flows.py

```diff
@@ -0,0 +1,278 @@
+import pandas as pd
+from bblocks import set_bblocks_data_path
+from bblocks.dataframe_tools.add import add_iso_codes_column
+from bblocks.import_tools.imf_weo import WEO
+import numpy as np
+
+from scripts.config import Paths
+
+set_bblocks_data_path(Paths.raw_data)
+
+
+def get_csv(doc: str) -> pd.DataFrame:
+    """
+    Loads csv from output folder based.
+
+    Args:
+        doc (str): Name of csv file.
+    """
+    return pd.read_csv(Paths.raw_data / doc)
+
+
+def get_parquet(doc: str) -> pd.DataFrame:
+    """
+    Loads parquet file from output folder based.
+
+    Args:
+        doc (str): Name of parquet file.
+    """
+    return pd.read_parquet(Paths.output / doc)
+
+
+def filter_for_specified_years(df: pd.DataFrame, years: list[int]) -> pd.DataFrame:
+    """
+    Filters the dataframe to keep only the rows for specified years.
+
+    Args:
+        df (pd.DataFrame)
+        years (list[int]): list of required years for filtered DataFrame.
+    """
+    return df.loc[lambda d: d.year.isin(years)]
+
+
+def calculate_net_transfers(df: pd.DataFrame) -> pd.DataFrame:
+    """
+    Creates net_flow column, which is the sum of inflows (+ve) and outflows (-ve).
+    Removes inflow and outflow columns for output.
+
+    Args:
+        df (pd.DataFrame)
+
+    Returns: pd.DataFrame with additional column for 'net_flow' and without inflow and
+    outflow columns. net_flow is presented in billions.
+    """
+    df["net_flow"] = df["inflow"] + df["outflow"]
+
+    df["net_flow"] = df["net_flow"] / 1000000000
+
+    return df.filter(
+        items=["year", "country", "continent", "income_level", "net_flow"], axis=1
+    )
+
+
+def add_latest_year_for_missing_gdp_data(
+    df: pd.DataFrame, gdp: pd.DataFrame
+) -> pd.DataFrame:
+    """
+    Fills gaps in GDP data where the year does not currently have data. Fills gaps with
+    latest available year.
+
+    Args:
+        df (pd.DataFrame) with missing GDP data
+        gdp (pd.DataFrame) with all gdp data
+
+    Returns: pd.DataFrame with gdp data for the latest year available, where neccessary.
+    """
+
+    # Make a copy to avoid modifying the original DataFrame
+    df = df.copy()
+
+    # Identify rows with missing GDP data
+    missing_gdp_indices = df.loc[lambda d: d.gdp.isna()].index
+
+    # Loop through rows with missing GDP data and fill in the latest year available
+    for idx in missing_gdp_indices:
+        iso = df.loc[idx, "iso_3"]
+        year = df.loc[idx, "year"]
+
+        # Find the latest year before the current 'year' for which GDP data is available
+        past_years = gdp.loc[lambda d: (d["iso_3"] == iso) & (d["year"] < year)]
+        if not past_years.empty:
+            latest_year = past_years["year"].max()
+            latest_gdp = gdp.loc[
+                lambda d: (d["iso_3"] == iso) & (d["year"] == latest_year), "gdp"
+            ].iloc[0]
+            df.loc[idx, "gdp"] = latest_gdp
+
+    return df
```

Comment:
This is quite a complex approach when it doesn't need to be. You shouldn't need to loop, or rely on index, or have complex logic to figure this out.

Here's an alternative. It:
- It makes sure the df is well sorted
- It forward fills the missing data (grouping by country). 

```python
# Make sure data is sorted by year
    df = df.sort_values(by=["iso_3", "year"])

    # Forward fill the missing gdp values
    df["gdp"] = df.groupby(
        ["country", "continent", "income_level", "iso_3"], dropna=False
    )["gdp"].fillna(method="ffill")
    
    return df
```

Note in this version I have not set a limit in terms of how far forward it will take the latest not-empty value. That's ok in the context that we just want the most recent possible one. In other contexts we may want to limit that.

Reply:
Wow yeah... that is much simpler, thank you!